util.go

```diff
@@ -144,22 +146,34 @@ func (b *Builder) copyFileWithTar(tarIDMappingOptions *IDMappingOptions, chownOp
 		if err != nil {
 			return errors.Wrapf(err, "error generating tar header for: %q", src)
 		}
-		hdr.Name = filepath.Base(dest)
+		chrootedDest, err := filepath.Rel(b.MountPoint, dest)
+		if err != nil {
+			return errors.Wrapf(err, "error generating relative-to-chroot target name for %q", dest)
+		}
+		hdr.Name = chrootedDest
 		hdr.Uid = int(containerUID)
 		hdr.Gid = int(containerGID)
 
 		if fi.Mode().IsRegular() && hdr.Typeflag == tar.TypeReg {
-			f, err = os.Open(src)
-			if err != nil {
-				return errors.Wrapf(err, "error opening %q to copy its contents", src)
+			if linkname := hardlinkChecker.Check(fi); linkname != "" {
+				hdr.Typeflag = tar.TypeLink
+				hdr.Linkname = linkname
+			} else {
+				hardlinkChecker.Add(fi, chrootedDest)
 			}
-			defer func() {
-				if f != nil {
-					if err := f.Close(); err != nil {
-						logrus.Debugf("error closing %s: %v", f.Name(), err)
-					}
+			if hdr.Typeflag == tar.TypeReg {
```

Comment:
Is this check necessary, isnce you have checked it above?

Reply:
We might be changing it from TypeReg to TypeLink at line 159.

Reply:
Then why not put this in else block above.

Reply:
Good point, changing it.